### PR TITLE
Support build on tar file

### LINKED ISSUE #19: Add support for source provided as tar archive

### PR DESCRIPTION
**Pull request by shhsu**
This add support for building from a remote url that is a tar.gz archive using a`--archive` option. 

**Fixes**
Fixes #19 

@Azure/azure-container-registry

### CODE REVIEW

**Comment by shhsu** [comment]
@sajayantony @shiweiz

**Comment by shhsu** [comment]
@sajayantony Moved and renamed packages as requested, but generating the project tar file on the fly require decoupling ArchiveSource with http client. I will have to leave this to the future.

Also I will have to create a PR changing the test_config.json file after this one to move to use the dev branch's project archive as test resource

**Comment by sajayantony** [comment]
Please remove the tar from the test/resources and let the test archive it instead.

File: README.md [hunk 1]
```diff
@@ -31,14 +31,14 @@ On your project directory to build you project and push to desired registry.
 ##### Conditional or Optional parameters
 `--docker-registry` Docker registry to push to. This parameter will populate the `ACR_BUILD_DOCKER_REGISTRY` reserved environment variable (see `Build Environment`) Registry is required if `--push` options is present<br />
 `--docker-user` Username for the docker registry specified above<br />
-`--docker-secret` Password or token for registry specified above<br />
+`--docker-password` Password or token for registry specified above<br />
+`--web-archive` The URL of a tar.gz archive which contains the source code<br />
```
Comment (sajayantony) [inline README.md]: >web [](start = 3, length = 3)

Lets rename this to --archive. 
I'm guessing you should be able to just merge --git-url and this to determine if its a clonable or just downloadable

File: pkg/commands/archive.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// NewArchiveSource creates a new archive source
+func NewArchiveSource(url string, targetDir string) build.Source {
+	return &ArchiveSource{url: url, targetDir: targetDir}
+}
+
+// Obtain downloads and extract the source
+func (s *ArchiveSource) Obtain(runner build.Runner) error {
+	response, err := http.Get(s.url)
```
Comment (sajayantony) [inline pkg/commands/archive.go]: >http.Get(s.url) [](start = 18, length = 15)

Would be good to enable timing info and output steps of download and extraction with timing here.

File: pkg/commands/git.go [hunk 3]
```diff
@@ -6,11 +6,13 @@ import (
 	"strconv"
 	"strings"
 
+	build "github.com/Azure/acr-builder/pkg"
 	"github.com/Azure/acr-builder/pkg/constants"
-	"github.com/Azure/acr-builder/pkg/domain"
 	"github.com/sirupsen/logrus"
 )
 
+const defaultTargetDir = "acr-build-source" // force targetDir to not be null so we never clone against .
```
Comment (sajayantony) [inline pkg/commands/git.go]: >acr-build-source [](start = 26, length = 16)

lets just keep the default as src.

File: pkg/constants/etc.go [hunk 4]
```diff
@@ -10,3 +10,12 @@ const TimestampFormat = time.RFC3339
...
 const DefaultDockerfile = "Dockerfile"
+
+// SourceNameWebArchive is the name of the web archive source
+const SourceNameWebArchive = "web archive"
+
+// SourceNameGit is the name of the git source
+const SourceNameGit = "git repository"
+
+// SourceNameLocal is the name of local source
+const SourceNameLocal = "local directory"
```
Comment (sajayantony) [inline pkg/constants/etc.go]: Please open an issue to merge these and enable align with Docker build.

File: pkg/commands/archive.go (hunk 2 above)
Comment (shhsu) [reply pkg/commands/archive.go]: https://github.com/Azure/acr-builder/issues/22

File: pkg/constants/etc.go (hunk 4 above)
Comment (shhsu) [reply pkg/constants/etc.go]: https://github.com/Azure/acr-builder/issues/21

**Comment by sajayantony** [comment]
This is failing  - 
```
./acr-builder --archive https://github.com/Azure/acr-builder/archive/master.zip --docker-file tests/resources/docker-compose/hello-node/Dockerfile.alpine --docker-image test

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x110d2a2]

goroutine 1 [running]:
compress/gzip.(*Reader).Close(0x0, 0xc42030a240, 0xc420000180)
	/usr/local/go/1.9.2/go/src/compress/gzip/gunzip.go:292 +0x22
github.com/Azure/acr-builder/pkg/commands.unTAR.func1(0x0)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/commands/archive.go:83 +0x2f
github.com/Azure/acr-builder/pkg/commands.unTAR(0x1306e7b, 0x1, 0x1460000, 0xc42030a300, 0x145fb00, 0xc420040390)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/commands/archive.go:89 +0x516
github.com/Azure/acr-builder/pkg/commands.(*ArchiveSource).Obtain(0xc42007d110, 0x1464ee0, 0xc420040d60, 0x0, 0x0)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/commands/archive.go:48 +0x231
github.com/Azure/acr-builder/pkg.(Source).Obtain-fm(0x1464ee0, 0xc420040d60, 0x2f3a4b97, 0xed19e9d60)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/driver/build.go:217 +0x43
github.com/Azure/acr-builder/pkg/workflow.(*Workflow).ScheduleRun.func1(0x1464ee0, 0xc420040d60, 0xc42007d170, 0x1, 0x1463ac0)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/workflow/workflow.go:70 +0x39
github.com/Azure/acr-builder/pkg/workflow.(*Workflow).Run(0xc42007d170, 0x1464ee0, 0xc420040d60, 0x0, 0x0)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/workflow/workflow.go:49 +0xb5
github.com/Azure/acr-builder/pkg/driver.(*Builder).Run(0xc42004fee8, 0x1306e7d, 0x1, 0x0, 0x0, 0x0, 0x0, 0x7fff5fbff9f6, 0x3b, 0x7fff5fbffa41, ...)
	/Users/sajay/code/src/github.com/Azure/acr-builder/pkg/driver/build.go:67 +0x440
main.main()
	/Users/sajay/code/src/github.com/Azure/acr-builder/main.go:69 +0xd69

```

**Pull request merged**